File: core/aggregators/mein_mmo/aggregator.py
"""Mein-MMO aggregator implementation."""

from typing import Any, Dict, List, Optional, Tuple

from ..utils import clean_html, format_article_content, remove_image_by_url
from ..utils.youtube import proxy_youtube_embeds
from ..website import FullWebsiteAggregator
from .comment_extractor import extract_comments
from .content_extraction import extract_mein_mmo_content
from .multipage_handler import detect_pagination, fetch_all_pages


class MeinMmoAggregator(FullWebsiteAggregator):
    """Specialized aggregator for Mein-MMO.de gaming news."""

    MEIN_MMO_URL = "https://mein-mmo.de/"

    def __init__(self, feed):
        super().__init__(feed)
        # Comment source: fetch_all_pages returns only the combined
        # entry-content blocks, so the multi-page raw_content has no thread.
        self._first_page_html: Optional[str] = None
        # Use Mein-MMO RSS feed if identifier is not set
        if not self.identifier or self.identifier == "":
            self.identifier = "https://mein-mmo.de/feed/"

    def get_source_url(self) -> str:
        """Return the Mein-MMO website URL."""
        return self.MEIN_MMO_URL

    @classmethod
    def get_identifier_choices(
        cls, query: Optional[str] = None, user: Optional[Any] = None
    ) -> List[Tuple[str, str]]:
        """Get available Mein-MMO RSS feed choices."""
        return [
            ("https://mein-mmo.de/feed/", "Main Feed (All Articles)"),
        ]

    @classmethod
    def get_default_identifier(cls) -> str:
        """Get default Mein-MMO identifier."""
        return "https://mein-mmo.de/feed/"

    @classmethod
    def get_configuration_fields(cls) -> Dict[str, Any]:
        """Get Mein-MMO configuration fields."""
        from django import forms

        return {
            "combine_pages": forms.BooleanField(
                initial=True,
                label="Combine Multi-page Articles",
                help_text="Automatically fetch and combine all pages of a multi-page article into one.",
                required=False,
            ),
            "include_comments": forms.BooleanField(
                initial=True,
                label="Include Comments",
                help_text="Extract wpDiscuz reader comments from the article page.",
                required=False,
            ),
            "max_comments": forms.IntegerField(
                initial=5,
                label="Max Comments",
                help_text="Maximum number of comments to extract per article.",
                required=False,
                min_value=0,
                max_value=20,
            ),
        }

    # Body lives in one known container -- keep the first match, never union.
    uses_first_content_match = True

    # Mein-MMO specific selectors
    content_selectors = ["div.entry-content"]

    selectors_to_remove = [
        "div.wp-block-mmo-recirculation-box",
        "div.wp-block-mmo-hub-box",
        "div.reading-position-indicator-end",
        "label.toggle",
        "a.wp-block-mmo-content-box",
        "div.page-links",
        "div.sources-wrapper",
        "div.feedback-box",
        "div.wp-block-wbd-affiliate-widget",
        "script",
        "style",
        "iframe:not([src*='youtube.com']):not([src*='youtu.be'])",
        "noscript",
        ".dailymotion-embed-container",
    ]

    def fetch_article_content(self, url: str) -> str:
        """
        Fetch article content, handling multi-page articles.
        """
        self.logger.debug(f"[fetch_article_content] Starting for URL: {url}")

        # Check configuration
        combine_pages = self.feed.options.get("combine_pages", True)

        # Fetch first page to detect pagination
        self.logger.debug("[fetch_article_content] Fetching first page")
        first_page_html = super().fetch_article_content(url)
        self._first_page_html = first_page_html
        self.logger.debug(
            f"[fetch_article_content] First page fetched ({len(first_page_html)} bytes)"
        )

        if not combine_pages:
            self.logger.info(f"[fetch_article_content] Multi-page combination disabled for {url}")
            return first_page_html

        # Check if multi-page
        self.logger.debug("[fetch_article_content] Detecting pagination")
        page_numbers = detect_pagination(first_page_html, self.logger)

        if len(page_numbers) <= 1:
            # Single page article
            self.logger.info(f"[fetch_article_content] Single page article detected for {url}")
            return first_page_html

        # Multi-page article - fetch all pages
        self.logger.info(
            f"[fetch_article_content] Multi-page article detected: {len(page_numbers)} pages"
        )

        combined_html = fetch_all_pages(
            base_url=url,
            page_numbers=page_numbers,
            fetcher=lambda page_url: super(MeinMmoAggregator, self).fetch_article_content(page_url),
            logger=self.logger,
            first_page_html=first_page_html,
        )

        self.logger.debug(
            f"[fetch_article_content] Returning combined HTML ({len(combined_html)} bytes)"
        )
        return combined_html

    def extract_content(self, html: str, article: Dict[str, Any]) -> str:
        """Extract Mein-MMO specific content."""
        self.logger.debug(f"[extract_content] Starting for {article.get('identifier')}")
        result = extract_mein_mmo_content(
            html=html,
            article=article,
            selectors_to_remove=self.selectors_to_remove,
            logger=self.logger,
        )
        self.logger.debug(f"[extract_content] Completed, result size: {len(result)} bytes")
        return result

    def process_content(self, html: str, article: Dict[str, Any]) -> str:
        """Process Mein-MMO content with header image extraction."""
        self.logger.debug(f"[process_content] Starting for {article.get('identifier')}")

        from bs4 import BeautifulSoup

        # Parse HTML for processing
        soup = BeautifulSoup(html, "html.parser")

        # Proxy YouTube embeds
        proxy_youtube_embeds(soup)

        # Remove header image from content if it was extracted
        header_data = article.get("header_data")
        if header_data and header_data.image_url:
            self.logger.debug(
                f"[process_content] Removing header image from content: {header_data.image_url}"
            )
            remove_image_by_url(soup, header_data.image_url)

        # Convert back to string for cleaning
        html = str(soup)

        # Clean HTML
        self.logger.debug("[process_content] Cleaning HTML")
        cleaned = clean_html(html)

        # Determine header image URL for formatting
        header_image_url = header_data.image_ref if header_data else None

        # wpDiscuz comments live on the article page outside the content div.
        comments_html = None
        include_comments = self.feed.options.get("include_comments", True)
        max_comments = self.feed.options.get("max_comments", 5)

        if include_comments:
            try:
                comment_source = self._first_page_html or article.get("raw_content", "")
                if comment_source:
                    comments_html = extract_comments(
                        comment_source,
                        article["identifier"],
                        max_comments=max_comments,
                        logger=self.logger,
                    )
            except Exception as e:
                self.logger.warning(f"[process_content] Failed to extract comments: {e}")

        # Format with header (image only), comments, and footer
        self.logger.debug("[process_content] Formatting content with header image only")
        formatted = format_article_content(
            cleaned,
            title=article["name"],
            url=article["identifier"],
            header_image_url=header_image_url,
            comments_content=comments_html,
        )

        self.logger.info(f"[process_content] Completed, formatted size: {len(formatted)} bytes")
        return formatted

File: src/lib/aggregators/sites/mein_mmo/aggregator.ts
import * as cheerio from "cheerio";
import { FeedLike, RawArticle } from "../../base";
import { isSafeUrl } from "../../blocks/parser";
import { cleanHtml, removeImageByUrl } from "../../extract/clean";
import { formatArticleContent } from "../../extract/format";
import { getHeaderImageRef } from "../../header/context";
import { storeImageRefFromUrl } from "../../images/store";
import { FirstPageStash, fetchAllPages } from "../../multipage";
import { FullWebsiteAggregator, proxyYoutubeEmbeds } from "../../website";
import { YOUTUBE_IFRAME_KEEP_SELECTOR } from "../../embeds/youtube-url";
import { extractComments } from "./comments";
import { extractMeinMmoContent, MEIN_MMO_CONTENT_SELECTORS } from "./content";
import { buildPageUrl, detectPagination } from "./multipage";

// The multi-page combine step (`fetchArticleContent()` below) selects by
// `MEIN_MMO_CONTENT_SELECTORS` (`./content.ts`) -- deliberately *not*
// `this.getContentSelectors()` (which would fold in a feed's
// `content_selectors` override). `extractMeinMmoContent()` has never read
// that override -- it hardcodes the same pair and returns the html untouched
// when neither is found -- so combining by a different selector set would
// hand it joined html it silently fails to recognise on any overridden feed,
// skipping Dailymotion-block handling, ignoreSelectors removal, embed
// processing and the "Weiter geht es auf Seite N" pagination-marker strip
// for that feed. Unifying the two sites' fetch *loop* (`../../multipage.ts`)
// does not require them to agree on selectors; making Mein-MMO honour
// `content_selectors` for real is a separate change to
// `extractMeinMmoContent()`, with its own tests.

export class MeinMmoAggregator extends FullWebsiteAggregator {
  static MEIN_MMO_URL = "https://mein-mmo.de/";
  static brandSiteUrl = "https://mein-mmo.de/";

  static getSourceUrl(): string {
    return MeinMmoAggregator.MEIN_MMO_URL;
  }

  override getSourceUrl(): string {
    return MeinMmoAggregator.MEIN_MMO_URL;
  }

  /**
   * The article headline. `h1.entry-title` is this WordPress theme's normal
   * heading; `og:title` is the fallback for the rare template that omits it.
   * Called once per page fetch (this aggregator paginates -- see
   * `fetchArticleContent()` below), and `noteSourceTitle()`'s "sticky" rule
   * (see ./base) means a later page's miss cannot blank out an earlier page's
   * match.
   */
  protected override sourceTitleFrom($: cheerio.CheerioAPI): string | null {
    const heading = $("h1.entry-title").first().text().trim();
    if (heading) return heading;
    const og = $('meta[property="og:title"]').attr("content");
    return og?.trim() || null;
  }

  static getDefaultIdentifier(): string {
    return "https://mein-mmo.de/feed/";
  }

  static getIdentifierChoices(): Array<[string, string]> {
    return [["https://mein-mmo.de/feed/", "Main Feed (All Articles)"]];
  }

  static getConfigurationFields(): Record<string, unknown> {
    return {
      combine_pages: {
        type: "boolean",
        initial: true,
        label: "Combine Multi-page Articles",
        help_text: "Automatically fetch and combine all pages of a multi-page article into one.",
        required: false,
      },
      include_comments: {
        type: "boolean",
        initial: true,
        label: "Include Comments",
        help_text: "Extract wpDiscuz reader comments from the article page.",
        required: false,
      },
      max_comments: {
        type: "number",
        initial: 5,
        label: "Max Comments",
        help_text: "Maximum number of comments to extract per article.",
        required: false,
        min_value: 0,
        max_value: 20,
      },
      include_videos: {
        type: "boolean",
        initial: false,
        label: "Include Auto-Inserted Videos",
        help_text:
          "Keep the Dailymotion player Mein-MMO's CMS drops into article bodies. " +
          "Off by default: the video it plays is chosen by the CMS, not the author, " +
          "and is often unrelated to the article.",
        required: false,
      },
    };
  }

  usesFirstContentMatch = true;

  static contentSelectors = [...MEIN_MMO_CONTENT_SELECTORS];
  protected contentSelectors = [...MeinMmoAggregator.contentSelectors];

  static selectorsToRemove = [
    "div.wp-block-mmo-recirculation-box",
    "div.wp-block-mmo-hub-box",
    // The "Inhalt" table of contents. Mein-MMO generates it with the Fixed TOC
    // (`ftwp`) WordPress plugin, which injects the whole widget -- header,
    // trigger button and the nested <ol> of anchors -- inside an otherwise
    // empty `<p class="wp-block-paragraph">` in the article body, so it is
    // extracted as article content and rendered as a stray numbered list.
    // It is navigation for the website's own page, not text: on a multi-page
    // article most of its entries are absolute links to /2/, /3/ and so on,
    // which do not exist in the aggregated article at all.
    //
    // Matched by id, and *only* this one: `div#ftwp-postcontent` is the same
    // plugin's wrapper around the ENTIRE article body, so a broader
    // `[id^='ftwp']` here would delete the article. The `<p>` left empty by
    // the removal is cleaned up by extractMeinMmoContent()'s
    // removeEmptyElements() pass.
    "div#ftwp-container-outer",
    "div.reading-position-indicator-end",
    "label.toggle",
    "a.wp-block-mmo-content-box",
    "div.page-links",
    "div.sources-wrapper",
    "div.feedback-box",
    "div.wp-block-wbd-affiliate-widget",
    "script",
    "style",
    YOUTUBE_IFRAME_KEEP_SELECTOR,
    "noscript",
    // Do NOT add ".dailymotion-embed-container" here! That is the facade
    // extractMeinMmoContent() builds for an author-inserted Dailymotion embed,
    // which is real article content. The CMS's own auto-inserted
    // "div.wp-block-mmo-video" blocks are a separate thing, and are dropped by
    // processDailymotionBlocks() when the feed's include_videos option is off
    // -- not from this list, so the removal can skip their thumbnail fetch too.
  ];
  protected selectorsToRemove = [...MeinMmoAggregator.selectorsToRemove];

  // Keyed by article URL, not a single field -- see `FirstPageStash`'s doc
  // comment in `../../multipage` for why: `enrichArticles()` runs
  // `fetchArticleContent()` for up to `this.concurrency` articles
  // concurrently on one aggregator instance, so a single field could be
  // overwritten by a sibling article's fetchArticleContent() while this
  // article's processContent() was still awaiting its img-resolution loop.
  // (This is the same shared stash MacTechNews now uses too -- see
  // ../mactechnews/aggregator.ts -- rather than a second, hand-rolled copy
  // of this same map.)
  private firstPages = new FirstPageStash();

  constructor(feed: FeedLike) {
    super(feed);
    if (!this.identifier) {
      this.identifier = "https://mein-mmo.de/feed/";
    }
  }

  override async fetchArticleContent(url: string): Promise<string> {
    const options = (this.feed.options as Record<string, unknown> | null) || {};
    const combinePages = options.combine_pages !== false;

    const firstPageHtml = await super.fetchArticleContent(url);
    this.firstPages.set(url, firstPageHtml);

    if (!combinePages) {
      return firstPageHtml;
    }

    const pageNumbers = detectPagination(firstPageHtml);
    if (pageNumbers.size <= 1) {
      return firstPageHtml;
    }

    const { combined } = await fetchAllPages(
      url,
      pageNumbers,
      MEIN_MMO_CONTENT_SELECTORS,
      (pageUrl) => super.fetchArticleContent(pageUrl),
      firstPageHtml,
      buildPageUrl,
    );

    return combined;
  }

  override async extractContent(html: string, article: RawArticle): Promise<string> {
    const labels = await this.chromeLabels();
    const options = (this.feed.options as Record<string, unknown> | null) || {};
    // `=== true`, not `!== false`, unlike combine_pages/include_comments above:
    // this option is off by default, so an absent value -- every feed created
    // before it existed -- must read as off.
    const includeVideos = options.include_videos === true;
    return extractMeinMmoContent(html, article, this.getIgnoreSelectors(), labels, includeVideos);
  }

  override async processContent(html: string, article: RawArticle): Promise<string> {
    const labels = await this.chromeLabels();
    const $ = cheerio.load(html);

    // Replace YouTube iframes with click-through facades
    await proxyYoutubeEmbeds($, labels);

    // Remove header image from content if extracted
    const headerData = article.header_data;
    if (headerData?.imageUrl) {
      removeImageByUrl($, headerData.imageUrl);
    }

    // Resolve body `<img>` sources to real yana-img:// references. A plain
    // for-of, not `.each()`: cheerio's `.each()` callback cannot be awaited,
    // and an article can carry more than one image.
    for (const imgEl of $("img").toArray()) {
      const $img = $(imgEl);
      const src = $img.attr("src");
      if (src && isSafeUrl(src)) {
        const ref = await storeImageRefFromUrl(src);
        if (ref) {
          $img.attr("src", ref);
        }
      }
    }

    const cleaned = cleanHtml($.html());
    const headerImageUrl = headerData ? getHeaderImageRef(headerData) : null;

    // Retrieve and clear this article's entry unconditionally -- fetchArticleContent()
    // always records one regardless of the include_comments option, so leaving the read
    // gated behind that option would leak an entry per article on every run with
    // comments disabled.
    const firstPageHtml = this.firstPages.take(article.identifier);

    let commentsHtml: string | null = null;
    const options = (this.feed.options as Record<string, unknown> | null) || {};
    const includeComments = options.include_comments !== false;
    const maxComments = typeof options.max_comments === "number" ? options.max_comments : 5;

    if (includeComments) {
      const commentSource = firstPageHtml || article.raw_content || "";
      if (commentSource) {
        // extractComments() never throws -- a selector-extraction failure is
        // caught and logged by the shared buildCommentsSection() it delegates
        // to (see src/lib/aggregators/comments/section.ts), rather than being
        // swallowed silently by a try/catch here.
        commentsHtml = extractComments(
          commentSource,
          article.identifier,
          maxComments,
          labels,
          this.onLog,
        );
      }
    }

    return formatArticleContent(
      cleaned,
      article.name,
      article.identifier,
      labels,
      headerImageUrl,
      null,
      commentsHtml,
    );
  }
}
